walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,67 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> options = new ArrayList<>();
+    List<Event> eventsSorted = new ArrayList<>();
+    eventsSorted.addAll(events);
+    Collections.sort(
+        eventsSorted,
+        new Comparator<Event>() {
+          @Override
+          public int compare(Event a, Event b) {
+            return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
+          }
+        });
+    eventsSorted.removeIf(
+        new Predicate<Event>() {
+          @Override
+          public boolean test(Event e) {
+            Collection<String> eventAttendees = new ArrayList<>();
+            eventAttendees.addAll(e.getAttendees());
+            eventAttendees.removeAll(request.getAttendees());
+            return e.getAttendees().size() == eventAttendees.size();
+          }
+        });
+
+    if (eventsSorted.isEmpty()) {
+      options.add(TimeRange.WHOLE_DAY);
+    } else {
+      Iterator<Event> eventsIter = eventsSorted.iterator();
+      Event event = eventsIter.next();
+      addTimeRangeIfPossible(
+          options, TimeRange.START_OF_DAY, event.getWhen().start(), request.getDuration());
+      while (eventsIter.hasNext()) {
+        Event nextEvent = eventsIter.next();
+        if (!nextEvent.getWhen().overlaps(event.getWhen())) {
+          addTimeRangeIfPossible(
+              options, event.getWhen().end(), nextEvent.getWhen().start(), request.getDuration());
+        }
+        event = nextEvent;
+      }
+      addTimeRangeIfPossible(
+          options, event.getWhen().end(), TimeRange.END_OF_DAY, request.getDuration());
+    }
+    if (TimeRange.WHOLE_DAY.duration() < request.getDuration()) {
+      options.clear();
+    }
+    return options;
+  }
```

Comment:
this could be another method that returns ```options``` and then is called to make the ```query()```

Reply:
I'll also refactor that out, thanks for pointing these out.